x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/processing_preview_table.tsx

```diff
@@ -0,0 +1,117 @@
+
+export function ProcessingPreviewTable({
+  selectedRowIndex,
+  onRowSelected,
+  showRowSourceAvatars,
+  originalSamples,
+  ...otherProps
+}: {
+  documents: SampleDocument[];
+  originalSamples?: SampleDocumentWithUIAttributes[];
+  displayColumns?: string[];
+  height?: EuiDataGridProps['height'];
+  renderCellValue?: (doc: SampleDocument, columnId: string) => React.ReactNode | undefined;
+  rowHeightsOptions?: EuiDataGridRowHeightsOptions;
+  toolbarVisibility?: boolean;
+  setVisibleColumns?: (visibleColumns: string[]) => void;
+  columnOrderHint?: string[];
+  sorting?: SimulationContext['previewColumnsSorting'];
+  setSorting?: (sorting: SimulationContext['previewColumnsSorting']) => void;
+  selectedRowIndex?: number;
+  onRowSelected?: (selectedRowIndex: number) => void;
+  showRowSourceAvatars: boolean;
+}) {
+  const leadingControlColumns: EuiDataGridControlColumn[] = useMemo(
+    () => [
+      {
+        id: 'selection',
+        width: showRowSourceAvatars ? 72 : 36,
+        headerCellRender: () => null,
+        rowCellRender: ({ rowIndex }) => {
+          const originalSample = originalSamples?.[rowIndex];
+          return (
+            <EuiFlexGroup gutterSize="s">
+              <EuiButtonIcon
+                onClick={() => {
+                  if (onRowSelected) {
+                    onRowSelected(rowIndex);
+                  }
+                }}
+                aria-label={i18n.translate(
+                  'xpack.streams.resultPanel.euiDataGrid.preview.selectRowAriaLabel',
+                  {
+                    defaultMessage: 'Select row {rowIndex}',
+                    values: { rowIndex: rowIndex + 1 },
+                  }
+                )}
+                iconType={selectedRowIndex === rowIndex ? 'minimize' : 'expand'}
+                color={selectedRowIndex === rowIndex ? 'primary' : 'text'}
+              />
+              {showRowSourceAvatars && originalSample && (
+                <RowSourceAvatar originalSample={originalSample} />
+              )}
+            </EuiFlexGroup>
+          );
+        },
+      },
+    ],
+    [onRowSelected, showRowSourceAvatars, selectedRowIndex, originalSamples]
+  );
+
+  return <PreviewTable {...otherProps} leadingControlColumns={leadingControlColumns} />;
+}
+
+const dataSourceColors = euiPaletteColorBlindBehindText();
+
+function dataSourceTypeToI18nKey(type: EnrichmentDataSourceWithUIAttributes['type']) {
+  switch (type) {
+    case 'random-samples':
+      return 'randomSamples';
+    case 'kql-samples':
+      return 'kqlDataSource';
+    case 'custom-samples':
+      return 'customSamples';
+  }
+}
+
+function RowSourceAvatar({ originalSample }: { originalSample: SampleDocumentWithUIAttributes }) {
+  const dataSource = useDataSourceSelector(
+    originalSample.dataSourceIndex,
+    (snapshot) => snapshot.context.dataSource
+  );
+  const color = dataSourceColors[originalSample.dataSourceIndex % dataSourceColors.length];
```

Comment:
This doesn't deterministically generate attributes for the documents as it relies on the order of the data sources. 
As suggested initially, it would be great to generate the data source attributes once and rely on them as a source of truth during the whole data source lifecycle, to guarantee visual consistency to the use when removing data sources not required anymore.
```ts
context: ({ input, self }) => ({
    parentRef: input.parentRef,
    dataSource: input.dataSource,
    streamName: input.streamName,
    data: [],
    uiAttributes: {
      color: dataSourceColors.shift(),
    }
  })

// Then here in the component usage
const color = useDataSourceSelector(
    originalSample.dataSourceId,
    (snapshot) => snapshot.context.uiAttributes.color
  );
```